services/StockService.php

```diff
@@ -214,43 +238,196 @@ public function AddProduct(int $productId, float $amount, $bestBeforeDate, $tran
+			}
+			
+			$amount = abs($amount - floatval($productDetails->stock_amount) - floatval($productDetails->product->tare_weight));
+		}
+
+		$checkNullLocation = $productDetails->product->location_id == $locationFromId;
+		$potentialStockEntries = $this->GetProductStockEntriesByLocation($productId, $locationFromId, $checkNullLocation);
+
+		if ($specificStockEntryId !== 'default')
+		{
+			$potentialStockEntries = FindAllObjectsInArrayByPropertyValue($potentialStockEntries, 'stock_id', $specificStockEntryId);
+		}
+
+		if ($checkNullLocation)
+		{
+			$productStockAmountByLocation = $this->Database->stock()->where('product_id = :1 AND (location_id = :2 OR location_id is null)', $productId, $locationFromId)->sum('amount');
+		} else {
+			$productStockAmountByLocation = $this->Database->stock()->where('product_id = :1 AND location_id = :2', $productId, $locationFromId)->sum('amount');
+		}
+		if ($amount > $productStockAmountByLocation)
+		{
+			throw new \Exception('Amount to be transfered cannot be > current stock amount');
+		}
+
+		foreach ($potentialStockEntries as $stockEntry)
+		{
+			if ($amount == 0)
+			{
+				break;
+			}
+			if ($amount >= $stockEntry->amount)
+			{
+				$logRow = $this->Database->stock_log()->createRow(array(
+					'product_id' => $productId,
+					'amount' => $stockEntry->amount * -1,
+					'best_before_date' => $stockEntry->best_before_date,
+					'purchased_date' => $stockEntry->purchased_date,
+					'stock_id' => $stockEntry->stock_id,
+					'transaction_type' => self::TRANSACTION_TYPE_TRANSFER_FROM,
+					'price' => $stockEntry->price,
+					'location_id' => $stockEntry->location_id
+				));
+		                $logRow->save();
+				//Add the amount into the new location
+				$this->AddProduct($productId, $stockEntry->amount, $stockEntry->best_before_date, self::TRANSACTION_TYPE_TRANSFER_TO, $stockEntry->purchased_date, $stockEntry->price, $locationId);
+
+				$amount -= $stockEntry->amount;
+				$stockEntry->delete();
+			} else {
+				$logRow = $this->Database->stock_log()->createRow(array(
+					'product_id' => $productId,
+					'amount' => $amount * -1,
+					'best_before_date' => $stockEntry->best_before_date,
+					'purchased_date' => $stockEntry->purchased_date,
+					'stock_id' => $stockEntry->stock_id,
+					'transaction_type' => self::TRANSACTION_TYPE_TRANSFER_FROM,
+					'price' => $stockEntry->price,
+					'location_id' => $stockEntry->location_id
+				));
+				$logRow->save();
+
+				//Add the amount into the new location
+				$this->AddProduct($productId, $amount, $stockEntry->best_before_date, self::TRANSACTION_TYPE_TRANSFER_TO, $stockEntry->purchased_date, $stockEntry->price, $locationId);
+
+				$stockEntry->update(array(
+					'amount' => $stockEntry->amount - $amount
+				));
+				$amount = 0;
+			}
+		}
+	}
+
 	public function ConsumeProduct(int $productId, float $amount, bool $spoiled, $transactionType, $specificStockEntryId = 'default', $recipeId = null)
```

Comment:
When it's now possible to consume a product from a specific location, how is this ensured when here no `location_id` is considered... (?)

Reply:
Ha, good point... I think I initially looked at it for specific stock entry and did not test without using that feature.